parser_library/src/lsp/lsp_context.cpp

```diff
@@ -0,0 +1,568 @@
+        return completion_list_s();
+    const text_data_ref_t& text = file_it->second->data;
+
+    char last_char =
+        (trigger_kind == completion_trigger_kind::trigger_character) ? trigger_char : text.get_character_before(pos);
+
+    if (last_char == '&')
+        return complete_var(file_it->second, pos);
+    else if (last_char == '.')
+        return complete_seq(file_it->second, pos);
+    else if (should_complete_instr(text, pos))
+        return complete_instr(file_it->second, pos);
+
+    return completion_list_s();
+}
+
+completion_list_s lsp_context::complete_var(const file_info_ptr& file, position pos) const
+{
+    auto scope = file->find_scope(pos);
+
+
+    completion_list_s items;
+    const vardef_storage& var_defs = scope ? scope->var_definitions : opencode_->variable_definitions;
+    for (const auto& vardef : var_defs)
+    {
+        auto cont = hover(vardef);
+        completion_item_s item(
+            "&" + *vardef.name, std::move(cont), "&" + *vardef.name, "", completion_item_kind::var_sym);
+        items.push_back(std::move(item));
+    }
+
+    return items;
+}
+
+completion_list_s lsp_context::complete_seq(const file_info_ptr& file, position pos) const
+{
+    auto macro_i = file->find_scope(pos);
+
+    const context::label_storage& seq_syms =
+        macro_i ? macro_i->macro_definition->labels : opencode_->hlasm_ctx.current_scope().sequence_symbols;
+
+    completion_list_s items;
+    for (const auto& sym : seq_syms)
+    {
+        std::string label = "." + *sym.second->name;
+        items.emplace_back(label, "Sequence symbol", label, "", completion_item_kind::seq_sym);
+    }
+    return items;
+}
+
+std::string get_macro_signature(const context::macro_definition& m)
+{
+    std::stringstream signature;
+    if (*m.get_label_param_name() != "")
+        signature << "&" << *m.get_label_param_name() << " ";
+    signature << *m.id << " ";
+
+    bool first = true;
+    const auto& pos_params = m.get_positional_params();
+    // First positional parameter is always label, even when empty
+    for (size_t i = 1; i < pos_params.size(); ++i)
+    {
+        if (pos_params[i] == nullptr)
+            continue;
+        if (!first)
+            signature << ",";
+        else
+            first = false;
+
+        signature << "&" << *pos_params[i]->id;
+    }
+    for (const auto& param : m.get_keyword_params())
+    {
+        if (!first)
+            signature << ",";
+        else
+            first = false;
+        signature << "&" << *param->id << "=" << param->default_data->get_value();
+    }
+    return signature.str();
+}
+
+
+bool is_comment(std::string_view line)
+{
+    return line.size() > 0 && (line[0] == '*' || (line.size() > 1 && line.substr(0, 2) == ".*"));
+}
+
+bool is_continued_line(std::string_view line)
+{
+    return line.size() > continuation_column && !isspace(line[continuation_column]);
+}
+
+
+std::string lsp_context::get_macro_documentation(const macro_info& m) const
```

Comment:
We are parsing comments that are around macro definition, so it can be shown in hover/completion. In the original implementation, we only showed 10 lines, but I don't think there is a good reason for that limitation. We also parse the macro signature.

Since some macros have the documentation before the signature, and some after, we just append both of them.

Reply:
there can be hundreds lines of changelogs too...